[MERGE FIRST] Porting Articles: Overview and Structure

cartermp opened the pull request:
Fixes #300.  Provides motivation for porting, an overview of the process, a quick list of tools you may use, and the file + structure for dealing with 3rd party dependencies.
- [x] Get MSTest story straight
- [x] Figure out a way to mention having .NET Core code coincide with .NET Framework code

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dotnet/core-docs/315)

<!-- Reviewable:end -->

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,72 @@
+   
+4. Port your tests over to xUnit and Nunit.
+
+   MSTest doesn't support .NET Core, so you'll have to use either xUnit or NUnit to test your .NET COre targets.  Because porting to .NET Core is such a big change to your codebase, it's highly recommended to get your tests ported so that you can run tests as you port projects over.
```
.NET Core, not COre

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,72 @@
+
+The recommended process for porting follows a series of steps.  Here's a short overview of the process in general.  There will be other specific steps to take when porting libraries, apps and services.
+
+1. Identify and account for your 3rd party dependencies.
```
maybe something to look at later. I think using third-party instead of 3rd party would be better.

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,72 @@
+   
+5. Determine which project system in Visual Studio is appropriate for your needs.
+
+   There are now two project systems in Visual Studio which support .NET Core: the traditional `csproj` project system and the new `xproj` project system.  Both offer capabilities that the other does not, so it's critical that you pick the correct project system before you begin.
```
How can you learn about the differences? Link somewhere?

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,72 @@
+
+If it makes sense to port at this time, then read on!
+
+## Overview of steps
```
This subtitle reads a bit strange to me. Maybe Porting process overview or 6-step porting process?

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,72 @@
+
+Here's a short list of the tools you'll likely use when porting to .NET Core:
+
+* Nuget - [client](https://dist.nuget.org/index.html) or [Nuget Package Explorer](https://github.com/NuGetPackageExplorer/NuGetPackageExplorer)
```
should we have a quick description of what they're for?

ghogen commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,72 @@
\ No newline at end of file
+By [Phillip Carter](https://github.com/cartermp)
+
+If you've got code running on .NET Framework you may be interested in running your code on .NET Core.  This article will cover evaluating if porting to .NET Core is reasonable for you, an overview of the porting process, and a list of the tools you may use to port.
```
add comma after .NET Framework

cartermp commented:
@bleroy I addressed feedback you left in another PR here.

cartermp commented:
@blackdwarf I addressed your two feedback items here.

cartermp commented:
@bleroy I added a new section about co-evolving your .NET Framework and .NET Core code.  PTAL!

bleroy commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,81 @@
+Imagine being able to continue using .NET Framework and Windows for your application, but being able to build and deploy different services you consume with the freedom that .NET Core offers.  A WPF GUI could consume services written in .NET Core, and the code for both the GUI and the services could live in the same Visual Studio solution.  This would allow a seamless debugging experience while a ...
+
+.NET Core also offers binary sharing, allowing you to write code once that is guaranteed to run on both your .NET Framework code and your .NET Core code.  This binary compatibility, which is covered by the [.NET Platform Standard](https://github.com/dotnet/corefx/blob/master/Documentation/architecture/net-platform-standard.md), allows you to continue to create components of your system in a modul ...
+
```
I like the new section.

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+
+## Aside: Co-evolving your .NET Core and .NET Framework code
+
+.NET Framework is the best way to build applications for the Windows platform.  As such, there's a good chance that you have significant amounts of code that is simply not applicable when considering porting to .NET Core.  And that's just fine!  The introduction of .NET Core opens up even more possibilities for your exist assets.
```
exist assets -> existing assets?

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+
+Here's a short list of the tools you'll likely use when porting to .NET Core:
+
+* NuGet - [client](https://dist.nuget.org/index.html) or [Nuget Package Explorer](https://github.com/NuGetPackageExplorer/NuGetPackageExplorer), the package manager for the .NET Platform.
```
NuGet Package Explorer with capital G

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
\ No newline at end of file
+By [Phillip Carter](https://github.com/cartermp)
+
+If you've got code running on .NET Framework you may be interested in running your code on .NET Core.  This article covers evaluating if porting to .NET Core is reasonable for you, an overview of the porting process, and a list of the tools you may use to port.
```
This sentence "porting to .NET Core is reasonable for you" reads a bit strange to me. Maybe it would read better if it said something like "porting to .NET Core is applicable to your project(s)" or maybe if you change reasonable to feasible.

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+
+## Should you port?
+
+The first thing you need to do is evaluate if it makes sense to port .NET Framework code over to .NET Core in the first place.  .NET Core offers some attractive capabilities, such as:
```
Extra space before the second sentence

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+
+The first thing you need to do is evaluate if it makes sense to port .NET Framework code over to .NET Core in the first place.  .NET Core offers some attractive capabilities, such as:
+
+- Using ASP.NET Core to write great web apps and services which you can write and deploy on OS X, supported Linux distributions, or Windows
```
why just ASP.NET here? should we be mentioning that .NET Core is cross plat in general?

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+
+Your code could be a good candidate for porting to .NET Core if...
+
+- You have an ASP.NET web app that you'd like to deploy to more than .NET Framework
```
I don't understand this sentence. What do you mean by deploy to more than .NET Framework?

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+- You're happy with your code running on .NET Framework as it is today
+- Your code depends heavily on COM interop
+
+If it makes sense to port now, then read on!
```
why now? Also read on sounds too informal to me, more appropriate for a blog post or something. Not sure what others think. Maybe something like "If it makes sense to port your code, then continue to read."

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+
+1. Identify and account for your third-party dependencies.
+
+   This will involve understanding what your 3rd party dependencies are, how you depend on them, how to see if they also run on .NET Core, and steps you can take if they don't.
```
3rd party still present in some places. Consider changing to third-party

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+
+   This ensures that you can use API alternatives for .NET Framework-specific targets in the cases where .NET Core can't support a particular API.
+   
+3. Use the [API Portability Analyzer](tooling.md) to analyze your assemblies and develop a plan based on its results.
```
"the analyzer to analyze" sounds repetitive. Maybe to examine?

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+   
+5. Determine which project system in Visual Studio is appropriate for your needs.
+
+   There are now two project systems in Visual Studio which support .NET Core: the traditional project system and the new `xproj` project system.  Both offer capabilities that the other does not, so it's critical that you pick the correct project system.  This will be covered in other articles.
```
I'm guessing the article doesn't exist yet for you to link to it, right?

mairaw commented:
Looks good except for some minor feedback I left

cartermp committed:
Add overview of the porting process

Also added a new file, Analyzing your third party dependencies.  The original TOC did not include this, but it is an essential part of the porting process.

cartermp commented:
Rebased in full.  This ready to merge?  /cc @richlander @blackdwarf @mairaw

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
\ No newline at end of file
+By [Phillip Carter](https://github.com/cartermp)
+
+If you've got code running on .NET Framework you may be interested in running your code on .NET Core.  This article covers evaluating if porting to .NET Core is feasible for you, an overview of the porting process, and a list of the tools you may use to port.
```
Nit: add comma after framework and remove comma after process

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+- You have a compelling business case to run code on more than just Windows
+- You want to take advantage of more flexible deployment
+- You have expertise on non-Windows platforms you wish to leverage
+- You're looking for an opportunity to implement an architectural change for your services (e.g. a monolithic service to microservices deployed to containers)
```
e.g. -> for example

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+
+Here's a short list of the tools you'll likely use when porting to .NET Core:
+
+* NuGet - [client](https://dist.nuget.org/index.html) or [NuGet Package Explorer](https://github.com/NuGetPackageExplorer/NuGetPackageExplorer), the package manager for the .NET Platform.
```
Nit: should we say NuGet client instead of just client?

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+Here's a short list of the tools you'll likely use when porting to .NET Core:
+
+* NuGet - [client](https://dist.nuget.org/index.html) or [NuGet Package Explorer](https://github.com/NuGetPackageExplorer/NuGetPackageExplorer), the package manager for the .NET Platform.
+* ApiPort - [command line tool](https://github.com/Microsoft/dotnet-apiport/releases) or [Visual Studio Extension](https://visualstudiogallery.msdn.microsoft.com/1177943e-cfb7-4822-a8a6-e56c7905292b), a toolchain that can generate a report of how portable your code is between .NET Framework and .NET Core, with an assembly-by-assembly breakdown of issues.
```
Previously in this topic you called this "API Portability Analyzer tool". Should the names match in both sections? Should you link to the tooling topic for more information too?

mairaw commented on docs/porting/overview.md:
```diff
@@ -1,5 +1,82 @@
+
+## Aside: Co-evolving your .NET Core and .NET Framework code
+
+.NET Framework is the best way to build applications for the Windows platform.  As such, there's a good chance that you have significant amounts of code that is simply not applicable when considering porting to .NET Core.  And that's just fine!  The introduction of .NET Core opens up even more possibilities for your existing assets.
```
Nit: The .NET Framework

mairaw commented:
I think it looks great. I left some really minor feedback

cartermp committed:
address mairaw feedback

cartermp committed:
address mairaw feedback

cartermp commented:
@mairaw addressed your feedback.

cartermp committed:
Address @richlander feedback and remove non-porting info

BillWagner commented:
@cartermp Should we merge this one? I can do the conflict resolution and merge, if that would help push this forward.

cartermp commented:
@BillWagner I was thinking about how to change these articles given the recent changes in direction both for tooling and overall API area.  I don't think there's enough concrete stuff to use such that these articles provided much value to people.

BillWagner commented:
@cartermp I was wondering the same over the weekend when I was going through the TOC from the master branch.

@blackdwarf Thoughts? Is the content relevant, but needs to move? The guidance for porting is worthwhile, and some content already exists in master. Others are placeholders

Also, @cartermp I didn't review the other content that filled out some of the placeholders started by this PR. Do you have recommendations on the usefulness of those articles?

cartermp commented:
>  Do you have recommendations on the usefulness of those articles?

I think the one that *was* the most useful would have been the libraries article.  Much of that content is still true today, but is going to be outdated by the next version of VS.  It could be worth changing the wording to suggest everything is a point-in-time thing, and that these articles are for early adopters and not for "regular" developers.  At least that's my reasoning.

cartermp committed:
reword porting and remove project system details

cartermp commented:
This is ready for final review + merge.  I removed:

* Outdated section about MSTest not being supported on .NET Core
* Choice about project system.  Although this is still a decision to make, it is strictly a point-in-time one.
